lib/luahid.c

```diff
@@ -0,0 +1,102 @@
+/*
+* SPDX-FileCopyrightText: (c) 2025 Jieming Zhou <qrsikno@gmail.com>
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/slab.h>
+#include <linux/limits.h>
+#include <asm/byteorder.h>
+
+#include <linux/hid.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <lunatik.h>
+
+typedef struct luahid_s {
+	lunatik_object_t *runtime;
+	struct hid_driver driver;
+} luahid_t;
+
+static const struct hid_device_id hid_table[] = {
+	{HID_DEVICE(HID_BUS_ANY, HID_GROUP_ANY, HID_ANY_ID, HID_ANY_ID)},
+	{ }
+};
+MODULE_DEVICE_TABLE(hid, hid_table);
+
+static void luahid_release(void *private)
+{
+	luahid_t *hid = (luahid_t *)private;
+	if (hid) {
+		hid_unregister_driver(&hid->driver);
+		lunatik_putobject(hid->runtime);
+	}
+}
+
+static int luahid_register(lua_State *L);
+
+static const luaL_Reg luahid_lib[] = {
+	{"register", luahid_register},
+	{NULL, NULL}
+};
+
+static const luaL_Reg luahid_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luahid_class = {
+	.name = "hid",
+	.methods = luahid_mt,
+	.release = luahid_release,
+	.sleep = true,
+};
+
+static int luahid_register(lua_State *L)
+{
+	luaL_checktype(L, 1, LUA_TTABLE); /* assure that is a driver */
+
+	lunatik_object_t *object = lunatik_newobject(L, &luahid_class, sizeof(luahid_t));
+	luahid_t *hid = (luahid_t *)object->private;
+
+	/*
+	 * configure the driver's properties & callbacks
+	 */
+	struct hid_driver *user_driver = &(hid->driver);
+	user_driver->name = lunatik_checkalloc(L, NAME_MAX);
+	lunatik_setstring(L, 1, user_driver, name, NAME_MAX);
+	user_driver->id_table = hid_table;
+
+	lunatik_registerobject(L, 1, object);
+	lunatik_setruntime(L, hid, hid);
+	lunatik_getobject(hid->runtime);
```

Comment:
I believe the pattern is to getobject(runtime) before setruntime().. btw, you didn't assign `hid->runtime`.. did you try your code? I think it is broken..

Reply:
my bad on that.. i misread the code.. sorry..